### PR TITLE
Issue #9133: AvoidEscapedUnicodeCharacters should support '\s' escape symbol

### PR DESCRIPTION
**Pull request by nmancus1**
Issue #9133: AvoidEscapedUnicodeCharacters should support '\s' escape symbol
and Issue #9159

Diff Regression projects: https://gist.githubusercontent.com/nmancus1/931b74d45ae7ff3b68f379898f9c816e/raw/c8d3c4f19440789c4eaaa17786ebc98d120dd22c/projects-to-test-on.properties

Diff Regression config: https://gist.githubusercontent.com/nmancus1/895dff8ab31b89ef9623e98391767441/raw/4a83b19384168311d39053b43a06d640b0cdcd55/config.xml

In an effort to make `TEXT_BLOCK_LITERAL` content match that of an identical (after compilation) `STRING_LITERAL` content, I attempted to do the following:
```
        String literal = ast.getText();
        if (ast.getType() == TokenTypes.TEXT_BLOCK_CONTENT) {
            literal = "\"" 
                + CheckUtil.stripIndentAndInitialNewLineFromTextBlock(ast.getText())
                + "\"";
        }

```
However, pitest complained that when the `if` statement condition is mutated to `true`, like so:
```
        String literal = ast.getText();
        if (true) {
            literal = "\""
                + CheckUtil.stripIndentAndInitialNewLineFromTextBlock(ast.getText())
                + "\"";
        }
```
It made no difference in violations. Instead of making some impossible test, I simply removed the `if` statement and the additional quotes, since they do not matter, based on the following:
1. `CheckUtil.stripIndentAndInitialNewLineFromTextBlock()` has no effect on `STRING_LITERAL` content, since `STRING_LITERAL`s have no preceding whitespace and newline.
2. The additional quotes I was adding made no difference(even when allowIfAllCharactersEscaped = true), since they are escaped anyway.

~~note: commit https://github.com/checkstyle/checkstyle/commit/9266613df869595a26a10c43d1771c4687d53d5c is not under review in this PR, I have included it because this PR depends on it. We can merge this PR after  https://github.com/checkstyle/checkstyle/pull/9134 is merged.~~

### CODE REVIEW

**Comment by nmancus1** [comment]
Github, generate report

**Review by romani — APPROVED** [review]
ok to merge  after first commit is merged in separate PR and code is rebased

**Comment by nmancus1** [comment]
Rebased.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/avoidescapedunicodecharacters/InputAvoidEscapedUnicodeCharactersEscapedS.java
Comment (strkkk) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/avoidescapedunicodecharacters/InputAvoidEscapedUnicodeCharactersEscapedS.java:27]: `// violation` comments should be on lines 19 and 25, where violation is reported

**Review by strkkk — CHANGES_REQUESTED** [review]
one minor

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/avoidescapedunicodecharacters/InputAvoidEscapedUnicodeCharactersEscapedS.java
Comment (nmancus1) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/avoidescapedunicodecharacters/InputAvoidEscapedUnicodeCharactersEscapedS.java]: @strkkk it is not possible to add comments to those particular lines; they would become text block content. I have added comments to make that more clear in the test case.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java:366]: I have added this code to ensure that the compiled content of both the text block literal and traditional string literal match where they should.

**Comment by nmancus1** [comment]
I have found some nuances in the logic for this check, please hold off on review until I resolve these and CI is green.

**Comment by romani** [comment]
If possible to postpone issue resolving, it is better to postpone. We can always fix one issue in next issue/PR.

**Comment by nmancus1** [comment]
> If possible to postpone issue resolving, it is better to postpone.
All is now resolved, left detailed comment in code.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/avoidescapedunicodecharacters/InputAvoidEscapedUnicodeCharactersEscapedS.java
Comment (romani) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/avoidescapedunicodecharacters/InputAvoidEscapedUnicodeCharactersEscapedS.java]: It is ok to to have `// violation` as part of string value. All we need is visual marker that line has violation

Comment (nmancus1) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/avoidescapedunicodecharacters/InputAvoidEscapedUnicodeCharactersEscapedS.java]: @romani in the case of text blocks, the initial newline and triple quotes delimit the content. I cannot put a comment or text following the delimiter on the same line.  I could do a C style comment:
```
                String test = /* violation */"""
                        """;
```

Comment (nmancus1) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/avoidescapedunicodecharacters/InputAvoidEscapedUnicodeCharactersEscapedS.java]: I left a comment after each text block, and a C style comment on the same line.

**Review by strkkk — APPROVED** [review]

**Comment by nmancus1** [comment]
Github, generate report

**Comment by nmancus1** [comment]
>I have found some nuances in the logic for this check, please hold off on review until I resolve these and CI is green.

New diff report shows nuance, namely that the check was trimming the first and last index of the `literal` string when checking for escape characters; this was only removing the `'` or `"` from the ends of a char literal or string literal, and does matter if we are just looking for escape characters. But, in the case of text blocks, they have no `"` to trim, so the check was removing characters at the beginning/end of the string(token text). 
Also, without stripping the text block string of the indent/initial newline, check was giving false positives for strings that were actually all unicode chars.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java:357]: I think we should move this text to PR description. This comment is hard to understand

Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java:357]: I put it where it is so that when someone reads the code later, and sees that we call `CheckUtil.stripIndentAndInitialNewLineFromTextBlock()` on all token text even if it is a `CHAR_LITERAL` or `STRING_LITERAL`, I want them to easily understand why this was done, since it seems a bit odd to use this method on text that is not from a text block...
I can change it if you disagree.

Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java:357]: I can do `// see https://github.com/checkstyle/checkstyle/pull/9135 for explanation` and put this text in the PR description.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java:195]: 1. Please change the link to https://docs.oracle.com/javase/specs/jls/se15/html/jls-3.html#jls-3.10.7

2. All patterns in this file seems to be sorted. It is better to place `\s` between `\r` and `\t`

3. The docs says there is one more escape - for a trailing `\`. Something like `\\$`.
For example
```java
    static final String value1 = """
        \u03bc\
        \s\u03bc\
        """; // ok
```
is just `μ μ`. Please create a new issue for this.

**Review by pbludov — CHANGES_REQUESTED** [review]
couple of notes

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java:195]: 1. Done
2. Done.
3. https://github.com/checkstyle/checkstyle/issues/9143

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java]: @romani I put a detailed comment in the PR description, and just did a 
`// See https://github.com/checkstyle/checkstyle/pull/9135#issue-545666710` comment before the assignment.

**Review by pbludov — APPROVED** [review]

**Comment by nmancus1** [comment]
Github, rebase

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java:354]: Confirm Issue https://github.com/checkstyle/checkstyle/issues/9159 represents this fix. I will approve that issue and you should split this into 2 commits, 1 for new functionality and 1 for a bug fix. Then we can merge it since others have all approved this.

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java:354]: Issue #9159 does represent this fix. PR split.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java:458]: Should this change be in the first commit?

**Review by rnveach — APPROVED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java:458]: Yes, apologies. Fixing now.

**Pull request merged**
